server/src/compression.rs

```diff
@@ -0,0 +1,473 @@
+    let mut compressed_segment_builder =
+        CompressedSegmentBuilder::new(uncompressed_timestamps, uncompressed_values, error_bound);
+
+    let mut current_index = start_index;
+    while compressed_segment_builder.can_fit_more() && current_index < end_index {
+        let timestamp = uncompressed_timestamps.value(current_index);
+        let value = uncompressed_values.value(current_index);
+        compressed_segment_builder.try_to_update_models(timestamp, value);
+        current_index += 1;
+    }
+
+    compressed_segment_builder
+}
+
+/// A compressed segment being build from an uncompressed segment using the
+/// model types in `models`.
+struct CompressedSegmentBuilder<'a> {
+    /// The regular timestamps of the uncompressed segment the compressed
+    /// segment is being build from.
+    uncompressed_timestamps: &'a TimestampArray,
+    /// The values of the uncompressed segment the compressed segment is being
+    /// build from.
+    uncompressed_values: &'a ValueArray,
+    /// Index of the first data point in `self.uncompressed_timestamps` and
+    /// `self.uncompressed_values` the compressed segment represents.
+    start_index: usize,
+    /// Constant function that currently represents the values in
+    /// `self.uncompressed_values` from `self.start_index` to `self.start_index`
+    /// + `self.pmc_mean.length`.
+    pmc_mean: PMCMean,
+    /// Indicates if `self.pmc_mean` could represent all values in
+    /// `self.uncompressed_values` from `self.start_index` to `current_index` in
+    /// `create_next_compressed_segment()`.
+    pmc_mean_has_fit_all: bool,
+    /// Linear function that represents the values in `self.uncompressed_values`
+    /// from `self.start_index` to `self.start_index` + `self.swing.length`.
+    swing: Swing,
+    /// Indicates if `self.swing` could represent all values in
+    /// `self.uncompressed_values` from `self.start_index` to `current_index` in
+    /// `create_next_compressed_segment()`.
+    swing_has_fit_all: bool,
+    /// Values in `self.uncompressed_values` from `self.start_index` to
+    /// `self.start_index` + `self.gorilla.length` compressed using lossless
+    /// compression.
+    gorilla: Gorilla,
+}
+
+impl<'a> CompressedSegmentBuilder<'a> {
+    fn new(
+        uncompressed_timestamps: &'a TimestampArray,
+        uncompressed_values: &'a ValueArray,
+        error_bound: ErrorBound,
+    ) -> Self {
+        Self {
+            uncompressed_timestamps,
+            uncompressed_values,
+            start_index: 0,
+            pmc_mean: PMCMean::new(error_bound),
+            pmc_mean_has_fit_all: true,
+            swing: Swing::new(error_bound),
+            swing_has_fit_all: true,
+            gorilla: Gorilla::new(),
+        }
+    }
+
+    /// Attempt to update the current models to also represent the `value` of
+    /// the data point collected at `timestamp`.
+    fn try_to_update_models(&mut self, timestamp: Timestamp, value: Value) {
```

Comment:
Maybe add a `debug_assert!` to this function that checks that `self.can_fit_more` is `true`.

Reply:
Fixed.